backends/ldap.go

```diff
@@ -0,0 +1,343 @@
+	SuperuserFilter          string
+	AclTopicPatternAttribute string
+	AclAccAttribute          string
+}
+
+func NewLDAP(authOpts map[string]string, logLevel log.Level) (LDAP, error) {
+
+	l, err := NewLDAPWithFactory(authOpts, logLevel, func(l LDAP) (LDAPClient, error) {
+		ldapClient, err := ldap.DialURL(l.Url)
+		return ldapClient, err
+	})
+
+	return l, err
+}
+
+func NewLDAPWithFactory(authOpts map[string]string, logLevel log.Level, ldapClientFactory LDAPClientFactory) (LDAP, error) {
+
+	log.SetLevel(logLevel)
+
+	ldapOk := true
+	missingOptions := ""
+
+	var l = LDAP{
+		LDAPClientFactory:        ldapClientFactory,
+		Url:                      "ldap://localhost:389",
+		GroupBaseDN:              "",
+		GroupFilter:              "(member=%s)",
+		SuperuserFilter:          "",
+		AclTopicPatternAttribute: "",
+		AclAccAttribute:          "",
+	}
+
+	if host, ok := authOpts["ldap_url"]; ok {
+		l.Url = host
+	}
+
+	if baseDN, ok := authOpts["ldap_base_dn"]; ok {
+		l.BaseDN = baseDN
+	} else {
+		ldapOk = false
+		missingOptions += " ldap_base_dn"
+	}
+
+	if groupBaseDN, ok := authOpts["ldap_group_base_dn"]; ok {
+		l.GroupBaseDN = groupBaseDN
+	}
+
+	if bindDN, ok := authOpts["ldap_bind_dn"]; ok {
+		l.BindDN = bindDN
+	} else {
+		ldapOk = false
+		missingOptions += " ldap_bind_dn"
+	}
+
+	if bindPass, ok := authOpts["ldap_bind_password"]; ok {
+		l.BindPass = bindPass
+	} else {
+		ldapOk = false
+		missingOptions += " ldap_bind_password"
+	}
+
+	if userFilter, ok := authOpts["ldap_user_filter"]; ok {
+		l.UserFilter = userFilter
+	} else {
+		ldapOk = false
+		missingOptions += " ldap_user_filter"
+	}
+
+	if groupFilter, ok := authOpts["ldap_group_filter"]; ok {
+		l.GroupFilter = groupFilter
+	}
+
+	if superuserFilter, ok := authOpts["ldap_superuser_filter"]; ok {
+		l.SuperuserFilter = superuserFilter
+	}
+
+	if aclTopicPatternAttribute, ok := authOpts["ldap_acl_topic_pattern_attribute"]; ok {
+		l.AclTopicPatternAttribute = aclTopicPatternAttribute
+	}
+
+	if aclAccAttribute, ok := authOpts["ldap_acl_acc_attribute"]; ok {
+		l.AclAccAttribute = aclAccAttribute
+	}
+
+	//Exit if any mandatory option is missing.
+	if !ldapOk {
+		return l, errors.Errorf("LDAP backend error: missing options:%s", missingOptions)
+	}
+
+	//Check if the LDAP server is reachable
+	ldapClient, err := l.LDAPClientFactory(l)
+	if err != nil {
+		log.Debugf("LDAP connection error: %s", err)
+		return l, err
+	}
+	l.LDAPClient = ldapClient
+
+	err = l.LDAPClient.Bind(l.BindDN, l.BindPass)
+	if err != nil {
+		log.Debugf("LDAP bind error: %s", err)
+		return l, err
+	}
+
+	return l, nil
+}
+
+func (l LDAP) GetUser(username, password, clientid string) (bool, error) {
+
+	searchRequest := ldap.NewSearchRequest(
+		l.BaseDN,
+		ldap.ScopeWholeSubtree,
+		ldap.NeverDerefAliases,
+		0,
+		0,
+		false,
+		fmt.Sprintf(l.UserFilter, ldap.EscapeFilter(username)),
+		[]string{"dn"},
+		nil,
+	)
+
+	searchResult, err := l.LDAPClient.Search(searchRequest)
```

Comment:
[Nitpick] Could you add empty lines between statements (not only here, but in general)?